Use Hottel's method for bifacial.utils.*_integ functions

kandersolar opened the pull request:
- [ ] Closes #xxxx
 - [ ] I am familiar with the [contributing guidelines](https://pvlib-python.readthedocs.io/en/latest/contributing.html)
 - [ ] Tests added
 - [ ] Updates entries in [`docs/sphinx/source/reference`](https://github.com/pvlib/pvlib-python/blob/main/docs/sphinx/source/reference) for API changes.
 - [ ] Adds description and name entries in the appropriate "what's new" file in [`docs/sphinx/source/whatsnew`](https://github.com/pvlib/pvlib-python/tree/main/docs/sphinx/source/whatsnew) for all changes. Includes link to the GitHub Issue with `` :issue:`num` `` or this Pull Request with `` :pull:`num` ``. Includes contributor name and/or GitHub username (link with `` :ghuser:`user` ``).
 - [ ] New code is fully documented. Includes [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) compliant docstrings, examples, and comments where necessary.
 - [ ] Pull request is nearly complete and ready for detailed review.
 - [ ] Maintainer: Appropriate GitHub Labels (including `remote-data`) and Milestone are assigned to the Pull Request and linked Issue.

I realized that our integrated view factor calculations are equivalent to what is calculated using Hottel's crossed strings method.  The calculations are straightforward, closed form, and much faster than numerical integration.  Here is a speed comparison for the ground->sky view factor:

![image](https://github.com/pvlib/pvlib-python/assets/57452607/ba8d5b3b-ec96-4c56-9c82-96ef81bca60a)

And since `vf_ground_sky_2d_integ` is the main bottleneck in `infinite_sheds.get_irradiance`, runtime improves there by 1-2 orders of magnitude as well.

PR is marked as draft while the code is a mess.

kandersolar committed:
draft of hottel's method in bifacial.utils.*_integ

mikofski commented:
FYI: there’s [another pvlib issue](https://github.com/pvlib/pvlib-python/issues/1488#issuecomment-1177842294) with links to recent uses of Hottel’s crossed strings in the literature, one of the main advantages being its east application for PV on slopes.

kandersolar commented:
> One disadvantage may be that we lose view factor as a function of location on the ground 

I don't see any reason not to retain the current `vf(x)` functions.  This PR only changes how the `_integ()` functions are computed internally.

kandersolar committed:
preliminary cleanup

kandersolar committed:
major refactor

kandersolar committed:
revert row_sky and row_ground to original definitions

the crossed-strings math assumed x0=0, x1=1.  I am confident it could be extended to other endpoints, but reverting this for now.

kandersolar committed:
more cleanup

kandersolar committed:
mimic original max_rows behavior (#1867)

kandersolar committed:
more max_rows behavior matching

kandersolar committed:
deprecate npoints and vectorize parameters

kandersolar committed:
more deprecation

kandersolar committed:
Merge remote-tracking branch 'upstream/main' into hottel

kandersolar committed:
replace npoints-based test with alternative

kandersolar committed:
lint

echedey-ls commented on pvlib/tests/bifacial/test_utils.py:
```diff
@@ -91,17 +93,23 @@ def test__vf_ground_sky_2d(test_system_fixed_tilt):
+    assert np.isclose(vf_expected, vf_actual)
+
+
+def test_vf_ground_sky_2d_integ_deprecated():
```
```suggestion
@fail_on_pvlib_version(...)
def test_vf_ground_sky_2d_integ_deprecated():
```
Let's make sure docs and code gets cleaned in the future? v0.12 or v0.13, whatever feels right. I would also add the future version to deprecate in the docs admonition.

kandersolar commented:
Closing as redundant with #2740
